Typescript, Wallbox, Batteries, Workflow Actions

mskg opened the pull request:
- Refactored to typescript
- Fixed battery charging states
- Added support for Alpha EV charger (requires Homey 12.4.5)

mskg committed:
- Refactored to typescript
- Added support for wallbox
- Fixed battery charging states

mskg committed:
Moved to Homey 12.4.5 evcharger capabilities

mskg commented:
You can test this version.

mskg committed:
- Refactored API devices to share statistic calls
- Added workflows for API battery, charger
- Added enery data for API grid, panels
- Fixed battery direction for API battery

mskg commented:
@Aboreon this needs retesting. 

- Refactored API devices to share statistic. This brings some relief to the Alpha APIs and also makes the data more consistent.
- Added workflows for API battery (ported from https://github.com/mskg/Homey-AlphaESS-inverter/pull/1), charger
- Added enery data for API grid, panels. They reset to 0 ever day - but Homey Energy is able to deal with that
- Fixed battery direction for API battery

mskg committed:
Fixed typo

Aboreon commented:
available as test now

Aboreon commented:
@mskg this latest version gets this error (about 2500 times now since in test with 10 testers)
Error: Invalid Capability: meter_power.imported
    at Remote Process
    at HomeyClient.emit (/node_modules/@athombv/homey-apps-sdk-v3/lib/HomeyClient.js:44:23)
    at Object.emit (/node_modules/@athombv/homey-apps-sdk-v3/manager/drivers.js:78:54)
    at Object.emit (/node_modules/@athombv/homey-apps-sdk-v3/lib/Driver.js:126:50)
    at /node_modules/@athombv/homey-apps-sdk-v3/lib/Device.js:568:27
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at GridDevice.setCapabilityValue (/node_modules/@athombv/homey-apps-sdk-v3/lib/Device.js:560:23)
    at GridDevice.handleEnergyData (/app/drivers/grid/device.js:16:18)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

mskg commented:
Understood. It seems to not create the new capabilities - for existing devices - as it should (automatically). Will check that.

mskg committed:
Fixed order of events. Capabilities must be updated before polling runs.

mskg commented:
@Aboreon  found it. The order of events was wrong. I added code to automatically account for missing capabilities. We added a few for every device. 

But, the necessary code-piece was run after the device initiated the initial polling. Such, the new properties didn't exist, while we already tried to fill values. 

Should be OK now.

Aboreon commented:
Homey.
I pulled your lastest changes via GitHub desktop.
Than ran the command homey app run and homey app validate both the same error

mskg committed:
Fixed charging control and states

mskg commented:
@Aboreon can you please check your "npm homey" version. I'm on 3.8.1. Please update `npm i -g homey`.
I checked the contents of the repro - everything is fine, and also the validate command succeeds. 

Independent of that: Alpha kept their word. They fixed the remote API, and I updated states and actions of the charger device. It now fully works.

Aboreon commented:
Fixed, on my side, its now in test.

Aboreon commented:
@mskg now a error is occuring with this stacktrace 108 in last 2 hours with 8 installs

Error: Invalid Capability: battery_charging_state
    at Remote Process
    at HomeyClient.emit (/node_modules/@athombv/homey-apps-sdk-v3/lib/HomeyClient.js:44:23)
    at Object.emit (/node_modules/@athombv/homey-apps-sdk-v3/manager/drivers.js:78:54)
    at Object.emit (/node_modules/@athombv/homey-apps-sdk-v3/lib/Driver.js:126:50)
    at /node_modules/@athombv/homey-apps-sdk-v3/lib/Device.js:568:27
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at BatteryDevice.setCapabilityValue (/node_modules/@athombv/homey-apps-sdk-v3/lib/Device.js:560:23)
    at BatteryDevice.setCapabilities (/app/drivers/battery_modbus/device.js:20:18)
    at ModbusEventEmitter.dataEmitter (/app/drivers/baseModbusDevice.js:77:24)
    at ModbusEventEmitter.emit (node:events:529:35)
    at ModbusEventEmitter.poll (/app/modbus/emitter.js:28:18)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

mskg committed:
Fixed modbus initialization

mskg commented:
The crash should be fixed; it had a wrong condition. It's now also the same code running for the non ModBus devices, maintaining the required capabilities. Did I get your question right?

For the second part: The flows are not available for the ModBus devices. So it's not not working, it's not available.

Aboreon commented:
ok, I put it in test! good for me to do next thing in another pr @mskg

mskg commented:
Than you can merge it down, and update the version# and release it. 🙌

mbnn commented:
Is it possible to add ev charger via modbus? 😀
For some strange reason this does not work: https://open.alphaess.com/login
When i click register and fill in my email i get no verification code nor can I login.

mskg commented:
Simply no. There is no possibility to do that via ModBus.

mbnn commented:
Ok but why am I not able to register? no verify code?

mskg commented:
We're not Alpha. Open a ticket via the App. Could take a bit, but they always reply / at least did in my cases.

mbnn commented:
Ok thanks, I will try 😀
